source/fuzz/fuzzer_pass_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,157 @@
+        // Instruction must be atomic or barrier instruction, return the number
+        // of memory semantic operands for instruction.
+        auto number_of_memory_semantics =
+            TransformationChangingMemorySemantics::GetNumberOfMemorySemantics(
+                inst_it->opcode());
+
+        // Equal zero if no atomic or barrier instruction is found.
+        if (number_of_memory_semantics == 0) {
+          return;
+        }
+
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfChangingMemorySemantics())) {
+          return;
+        }
+
+        // If the instruction has two memory semantic operands pick one
+        // randomly.
+        uint32_t memory_semantics_operand_position = 0;
+        if (number_of_memory_semantics == 2) {
+          std::vector<uint32_t> operand_positions{0, 1};
+          memory_semantics_operand_position =
+              GetFuzzerContext()->RandomIndex(operand_positions);
+        }
+
+        auto needed_index = TransformationChangingMemorySemantics::
+            GetMemorySemanticsOperandIndex(inst_it->opcode(),
+                                           memory_semantics_operand_position);
+
+        auto memory_semantics_value =
+            GetIRContext()
+                ->get_def_use_mgr()
+                ->GetDef(inst_it->GetSingleWordInOperand(needed_index))
+                ->GetSingleWordInOperand(0);
+
+        auto lower_bits_old_memory_semantics =
+            static_cast<SpvMemorySemanticsMask>(
+                memory_semantics_value & TransformationChangingMemorySemantics::
+                                             kMemorySemanticsLowerBitmask);
+        auto higher_bits_old_memory_semantics =
+            static_cast<SpvMemorySemanticsMask>(
+                memory_semantics_value & TransformationChangingMemorySemantics::
+                                             kMemorySemanticsHigherBitmask);
+
+        std::vector<SpvMemorySemanticsMask> memory_semanitcs_masks{
+            SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireMask,
+            SpvMemorySemanticsReleaseMask, SpvMemorySemanticsAcquireReleaseMask,
+            SpvMemorySemanticsSequentiallyConsistentMask};
+
+        // Remove the memory mask is not applicable for instruction.
+        memory_semanitcs_masks.erase(
+            std::remove_if(memory_semanitcs_masks.begin(),
+                           memory_semanitcs_masks.end(),
+                           [inst_it, lower_bits_old_memory_semantics](
+                               SpvMemorySemanticsMask /*unused*/) {
+                             switch (inst_it->opcode()) {
+                               case SpvOpAtomicLoad:
+                                 return TransformationChangingMemorySemantics::
+                                     IsAtomicLoadMemorySemanticsValue(
+                                         lower_bits_old_memory_semantics);
```

Comment:
Yes, you are right, but here I'm don't need `SpvMemorySemanticsMask /*unused*/` because I would like to filter the `potential_new_memory_orders` vector and remove elements that are not suitable for specific atomic or barrier instruction.
What do you think @paulthomson?

Reply:
Yes, you are right, but here I'm don't need `SpvMemorySemanticsMask /*unused*/` because I would like to filter the `potential_new_memory_orders` vector and remove elements that are not suitable for specific atomic or barrier instruction depend on `lower_bits_old_memory_semantics`.
What do you think @paulthomson?

Reply:
Let's discuss this via a video call.

Reply:
Could you please add a static function similar to:

```c++
std::vector<SpvMemorySemanticsMask> FuzzerPassChangingMemorySemantics::GetSuitableNewMemorySemanticsLowerBitValues(SpvOp opcode, SpvMemorySemanticsMask lower_bits_old_memory_semantics_value)
```

It should return a vector containing only the suitable new memory semantic values (lower bits) given the opcode and old memory semantics value (lower bits). E.g. for OpAtomicLoad and None, it should return [Acquire, SequentiallyConsistent]. You can then write a test for this function in a file called "fuzzer_pass_changing_memory_semantics_test.cpp".

You may end up adding other parameters to this function at some point, which is fine.

Reply:
Done.